pkg/broadcaster/broadcaster.go

```diff
@@ -0,0 +1,77 @@
+package broadcaster
+
+import (
+	"errors"
+	"io"
+	"sync"
+)
+
+// Broadcaster is a simplified Broker allowing clients to subscribe while
+// itself behaving as a io.Writer.
+type Broadcaster struct {
+	mu sync.Mutex
+	// NOTE: As we use a map, the broadcast order is randomized.
+	// Consider using a slice instead to be deterministic.
+	clients map[io.WriteCloser]struct{}
+}
+
+func NewBroadcaster() *Broadcaster {
+	return &Broadcaster{
+		clients: map[io.WriteCloser]struct{}{},
+	}
+}
+
+// Surface the lock as Pause/Unpause to freeze the broadcast while we make a copy of historical data.
+func (b *Broadcaster) Pause()   { b.mu.Lock() }
+func (b *Broadcaster) UnPause() { b.mu.Unlock() }
+
+// If broadcaster is closed, no nothing.
+func (b *Broadcaster) Subscribe(w io.WriteCloser) {
+	b.mu.Lock()
+	b.UnsafeSubscribe(w)
+	b.mu.Unlock()
+}
+
+// Needs to be gated by Pause/Unpause.
+func (b *Broadcaster) UnsafeSubscribe(w io.WriteCloser) {
+	if b.clients == nil {
+		return
+	}
+	b.clients[w] = struct{}{}
+}
+
+func (b *Broadcaster) Unsubscribe(w io.WriteCloser) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	if b.clients == nil {
+		return
+	}
+	delete(b.clients, w)
+}
+
+// NOTE: This approach is not ideal, if any client is slow or blocking,
+// it will slow/block everyone.
```

Comment:
This has the potential to actually block the command itself once the OS pipe it's using is full, right? Not just other readers.

Reply:
Correct. If the broadcaster is blocked everything is blocked. I'll add a mention of the command in the comment.